app/controllers/course/learning_map_controller.rb

```diff
@@ -1,15 +1,214 @@
+    conditional = get_conditional(parent_and_node_id_pair_params[:node_id])
+    condition.conditional = conditional
+
+    if condition.save
+      index
+    else
+      error_response(condition.errors.full_messages)
+    end
+  end
+
+  def remove_parent_node
+    condition = get_condition(parent_and_node_id_pair_params[:parent_node_id],
+                              parent_and_node_id_pair_params[:node_id])
+
+    if condition.destroy
+      index
+    else
+      error_response(condition.errors.full_messages)
+    end
+  end
+
+  def toggle_satisfiability_type
+    conditional = get_conditional(node_params[:node_id])
+
+    if conditional.satisfiability_type.to_s == :all_conditions.to_s
+      conditional.set_at_least_one_condition_satisfiability_type!
+    else
+      conditional.set_all_conditions_satisfiability_type!
+    end
+
+    if conditional.save
+      index
+    else
+      error_response(conditional.errors.full_messages)
+    end
+  end
 
   # @return [Course::LearningMapComponent]
   # @return [nil] If component is disabled.
   def component
     current_component_host[:course_learning_map_component]
   end
+
+  private
+
+  def error_response(errors)
+    respond_to do |format|
+      format.html
+      format.json do
+        render json: { errors: errors }, status: :bad_request
+      end
+    end
+  end
+
+  def init_conditionals
+    @conditionals = Course::Condition.conditionals_for(current_course)
+  end
+
+  def map_conditionals_to_nodes
+    all_node_relations = generate_all_node_relations
+    nodes = generate_nodes_from_conditionals(all_node_relations)
+    generate_node_depths(nodes)
+  end
+
+  def generate_all_node_relations
+    relations = init_all_node_relations
+    node_ids_to_children = relations[:node_ids_to_children]
+    node_ids_to_parents = relations[:node_ids_to_parents]
+
+    @conditionals.each do |conditional|
+      node_id = get_node_id(conditional)
+
+      conditional.conditions.each do |condition|
+        next if condition.actable_type == Course::Condition::Level.name
+
+        parent = map_condition_to_parent(condition)
+        node_ids_to_children[parent[:id]].push({ id: node_id, is_satisfied: parent[:is_satisfied] })
+        node_ids_to_parents[node_id].push(parent)
+      end
+    end
+
+    { node_ids_to_children: node_ids_to_children, node_ids_to_parents: node_ids_to_parents }
+  end
+
+  def init_all_node_relations
+    { node_ids_to_children: @conditionals.map { |conditional| [get_node_id(conditional), []] }.to_h,
+      node_ids_to_parents: @conditionals.map { |conditional| [get_node_id(conditional), []] }.to_h }
+  end
+
+  def map_condition_to_parent(condition)
+    type = condition.actable_type.demodulize
+    typed_condition = Object.const_get("Course::Condition::#{type}").find(condition.actable_id)
+    id = "#{type.downcase}-#{typed_condition.send("#{type.downcase}_id")}"
+
+    { id: id, is_satisfied: typed_condition.satisfied_by?(current_course_user) }
```

Comment:
Not sure if it'd be good to preload the stuff needed for the satisfiability check as well.

Reply:
Added preloading for the `actable` type for finding typed conditions.